### PR TITLE
View license obligations

### PR DESCRIPTION
**Pull request by AnnieJovitha**
Back end support to view License Obligations:
- returns map of obligations to Dependency

### CODE REVIEW

File: src/test/java/com/philips/research/bombar/core/domain/ProjectInteractorTest.java [hunk 1]
```diff
@@ -210,6 +210,30 @@ void readsLicenseDistributionForProject() {
...
+        @Test
+        void obligationsInAProject() {
+            ProjectService mockInteractor = spy(interactor);
+            when(store.getProject(PROJECT_ID)).thenReturn(Optional.of(project));
+            final var dependencies = Set.of(dependency.setLicense("LICENSE"));
+            final var obligations = new HashMap<String, Set<Dependency>>();
+            obligations.put("Obligation", dependencies);
+            final var dependencyDto = DtoConverter.toBaseDto(dependency);
+
+            doReturn(obligationAnalyzer).when(mockInteractor).getObligationAnalyzerInstance(PROJECT_ID);
```
Comment (timovandeput) [inline src/test/java/com/philips/research/bombar/core/domain/ProjectInteractorTest.java:229]: The code declares a class member for the mock analyzer, but this is the only location where that object is used.

Consider moving creation of this mock into this method (and simply name it `analyzer`).

File: src/test/java/com/philips/research/bombar/core/domain/ProjectInteractorTest.java [hunk 2]
```diff
@@ -210,6 +210,30 @@ void readsLicenseDistributionForProject() {
...
+            when(obligationAnalyzer.findObligations()).thenReturn(obligations);
+
+            final var obligationsResult = mockInteractor.getObligations(PROJECT_ID);
+            assertThat(obligationsResult.size()).isEqualTo(1);
```
Comment (timovandeput) [inline src/test/java/com/philips/research/bombar/core/domain/ProjectInteractorTest.java:233]: Not sure why you first check if there is one entry, and then access the entry.

Consider removing this line.

File: src/test/java/com/philips/research/bombar/core/domain/licenses/ObligationAnalyzerTest.java [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+        final var dependency = new Dependency(KEY, TITLE).setLicense(LICENSE);
+        project.addDependency(dependency);
+
+        final var obligations = analyzer.findObligations();
+
+        assertThat(obligations).isEqualTo(Map.of(OBLIGATION_DESC, Set.of(dependency)));
+    }
+
+    @Test
+    void obligationForMultiLicenses() {
```
Comment (timovandeput) [inline src/test/java/com/philips/research/bombar/core/domain/licenses/ObligationAnalyzerTest.java:63]: Reading the test, this appears to test "obligations for multi-license".

Consider renaming.

File: src/test/java/com/philips/research/bombar/core/domain/licenses/ObligationAnalyzerTest.java [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+        final var dependency = new Dependency(KEY, TITLE).setLicense(LICENSE + " AND " + LICENSE);
+        project.addDependency(dependency);
+
+        final var obligations = analyzer.findObligations();
+
+        assertThat(obligations).isEqualTo(Map.of(OBLIGATION_DESC, Set.of(dependency)));
+    }
+
+    @Test
+    void multipleObligationForLicense() {
```
Comment (timovandeput) [inline src/test/java/com/philips/research/bombar/core/domain/licenses/ObligationAnalyzerTest.java:74]: Suggestion to make explicit this is about a "single" license.

File: src/test/java/com/philips/research/bombar/core/domain/licenses/ObligationAnalyzerTest.java [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+
+        final var obligations = analyzer.findObligations();
+
+        assertThat(obligations).isEqualTo(Map.of("Unknown Obligations", Set.of(dependency)));
+    }
+
+    @Test
+    void obligationForLicense() {
+        registry.license(LICENSE).requires(OBLIGATION);
+        final var dependency = new Dependency(KEY, TITLE).setLicense(LICENSE);
```
Comment (timovandeput) [inline src/test/java/com/philips/research/bombar/core/domain/licenses/ObligationAnalyzerTest.java:54]: Creating a dependency and adding it to the project is repeated in most tests now.

Consider grouping these tests, and settings this up only once. (An alternative could be to at least declare and define a `private final Dependency dependency = new Dependency(KEY, TITLE)` at the top of the class.)

File: src/test/java/com/philips/research/bombar/core/domain/licenses/ObligationAnalyzerTest.java [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+        final var dependency1 = new Dependency("DEP1", TITLE).setLicense(LICENSE);
+        project.addDependency(dependency1);
+        final var dependency2 = new Dependency("DEP2", TITLE).setLicense(LICENSE);
+        project.addDependency(dependency2);
+
+        final var obligations = analyzer.findObligations();
+
+        assertThat(obligations).isEqualTo(Map.of(OBLIGATION_DESC, Set.of(dependency1, dependency2)));
+    }
+
```
Comment (timovandeput) [inline src/test/java/com/philips/research/bombar/core/domain/licenses/ObligationAnalyzerTest.java:122]: Two bonus blank lines? :-)

**Review by timovandeput — CHANGES_REQUESTED** [review]

File: src/main/java/com/philips/research/bombar/core/ProjectService.java [hunk 7]
```diff
@@ -86,6 +90,22 @@
...
+     */
+    Map<String, Set<DependencyDto>> getObligations(UUID projectId);
+
+    /**
+     * Returns an instance Obligation Analyzer for the project
+     *
+     * @param projectId target project
+     * @return instance of ObligationAnalyzer
+     */
+    ObligationAnalyzer getObligationAnalyzerInstance(UUID projectId);
```
[the remark this answers is not in the document]
Comment (AnnieJovitha) [reply src/main/java/com/philips/research/bombar/core/ProjectService.java:105]: If I make it as a private method, I will not be able to mock it , again I need to create this method in the interface just to mock it.

File: src/test/java/com/philips/research/bombar/core/domain/DependencyTest.java [hunk 8]
```diff
@@ -146,6 +146,35 @@ void addsUsages() {
...
+        final var staticRelation = new Relation(Relation.Relationship.STATIC_LINK, dependency);
+        final var dynamicRelation = new Relation(Relation.Relationship.DYNAMIC_LINK, dependency);
+
+        parent.addRelation(staticRelation);
+        parent.addRelation(dynamicRelation);
+        dependency.addUsage(parent);
+
+        assertThat(dependency.getStrongUsage()).contains(Relation.Relationship.STATIC_LINK);
+    }
+
```
[the remark this answers is not in the document]
Comment (AnnieJovitha) [reply src/test/java/com/philips/research/bombar/core/domain/DependencyTest.java:177]: sorry I dont get it

[the remark this answers is not in the document]
Comment (timovandeput) [reply src/test/java/com/philips/research/bombar/core/domain/DependencyTest.java:177]: There currently is no test for the situation where a single parent has multiple children. In that case the code should filter the relationships to the child dependency asking for it strongest relationship. If other children have an even stronger relationship, this should not impact the asking child dependency.

PS: This is the problem with test-after programming; detecting the full set of test cases from the code frequently is not trivial. In the test-first approach, that special handling could only have been added if a test confirmed it was still missing.

File: src/main/java/com/philips/research/bombar/core/domain/licenses/ObligationsAnalyzer.java [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+    }
+
+    /**
+     * @return map of license obligations mapped to the present dependencies including the license
+     */
+    public Map<String, Set<Dependency>> findObligations() {
+        Map<String, Set<Dependency>> obligations = new HashMap<>();
+        project.getDependencies().forEach(dep -> {
+            final var relationship = dep.getStrongestUsage().orElse(Relation.Relationship.weakest());
+            final var distribution = project.getDistribution();
```
Comment (timovandeput) [inline src/main/java/com/philips/research/bombar/core/domain/licenses/ObligationsAnalyzer.java:31]: This variable is now again defined for every dependency, although it is not related to the dependency.

My suggestion is to move the variable outside of the `forEach(...)` iteration.

File: src/main/java/com/philips/research/bombar/core/domain/licenses/ObligationsAnalyzer.java [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+    private final LicenseRegistry registry;
+    private final Project project;
+
+    public ObligationsAnalyzer(LicenseRegistry registry, Project project) {
+        this.registry = registry;
+        this.project = project;
+    }
+
+    /**
+     * @return map of license obligations mapped to the present dependencies including the license
```
Comment (timovandeput) [inline src/main/java/com/philips/research/bombar/core/domain/licenses/ObligationsAnalyzer.java:25]: I don't understand the "including the license" part. How is that relevant as the return value?

Suggestion to change the documentation to e.g. "@return map containing all dependencies per obligation".

You could additionally add a description for the method to explain that it derives obligations per license for each dependency. (See the common conventions for JavaDoc comments.)

File: src/test/java/com/philips/research/bombar/core/domain/DependencyTest.java [hunk 11]
```diff
@@ -146,6 +146,29 @@ void addsUsages() {
...
+    @Test
+    void noStrongUsageForOrphan() {
+        assertThat(dependency.getStrongestUsage()).isEmpty();
+    }
+
+    @Test
+    void findsStrongestRelationShipFromOwnUsage() {
+        final var parent = new Dependency(ID, "Parent");
+        final var firstChild = new Dependency(ID, "First");
+        final var secondChild = new Dependency(ID, "Second");
```
Comment (timovandeput) [inline src/test/java/com/philips/research/bombar/core/domain/DependencyTest.java:158]: This test is now testing two situations:
1. Multiple incoming relationships, where it should pick the strongest.
2. A sibling dependency that should be ignored.

Unit tests should only test one thing at a time. (This makes sure maintainers understand what the test tries to prove, and prevent accidentally passing one of the scenarios due to the more complex context.)

Note that the latter is now already happening: The second child relationship is is ignored because it is weaker than the other two relationships. (Proof: Removing line 159 of Dependency.java does not make this test fail.)

Split the two into separate tests.

File: src/test/java/com/philips/research/bombar/core/domain/ProjectInteractorTest.java [hunk 12]
```diff
@@ -210,6 +208,26 @@ void readsLicenseDistributionForProject() {
...
+        void noObligationsForEmptyProject() {
+            project.clearDependencies();
+
+            assertThat(interactor.findObligations(PROJECT_ID)).isEmpty();
+        }
+
+        @Test
+        void readObligationsForProject() {
+            final var analyzer = mock(ObligationsAnalyzer.class);
+            final var obligation = "Obligation";
```
Comment (timovandeput) [inline src/test/java/com/philips/research/bombar/core/domain/ProjectInteractorTest.java:221]: This looks like a candidate for extraction as an `OBLIGATION` constant.

File: src/test/java/com/philips/research/bombar/core/domain/licenses/ObligationsAnalyzerTest.java [hunk 13]
```diff
@@ -0,0 +1,123 @@
...
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class ObligationsAnalyzerTest {
+    private static final String OTHER_OBLIGATION = "Other";
+    private static final String OTHER_OBLIGATION_DESC = "Other desc";
+    private static final String KEY = "Key";
+    private static final String TITLE = "Title";
+    private static final String LICENSE = "License";
+    private static final String OBLIGATION = "Obligation";
+    private static final String OBLIGATION_DESC = "Obligation desc";
```
Comment (timovandeput) [inline src/test/java/com/philips/research/bombar/core/domain/licenses/ObligationsAnalyzerTest.java:22]: Could refactor this into a constant `DESCRIPTION = "Description"`, as the "description" aspect of this constant is more important than the "obligation" aspect.

File: src/test/java/com/philips/research/bombar/core/domain/licenses/ObligationsAnalyzerTest.java [hunk 14]
```diff
@@ -0,0 +1,123 @@
...
+import org.junit.jupiter.api.Test;
+
+import java.util.Map;
+import java.util.Set;
+import java.util.UUID;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class ObligationsAnalyzerTest {
+    private static final String OTHER_OBLIGATION = "Other";
```
Comment (timovandeput) [inline src/test/java/com/philips/research/bombar/core/domain/licenses/ObligationsAnalyzerTest.java:16]: Seems odd to first encounter the "other" obligation, and then later find the "main" obligation on line 21.

Always sort the constants to make it easier to recognise which constants are available.

**Review by timovandeput — CHANGES_REQUESTED** [review]

File: src/test/java/com/philips/research/bombar/core/domain/ProjectInteractorTest.java (hunk 12 above)
Comment (AnnieJovitha) [reply src/test/java/com/philips/research/bombar/core/domain/ProjectInteractorTest.java:221]: this constant is used only for this test, why do you want to extract?

Comment (timovandeput) [reply src/test/java/com/philips/research/bombar/core/domain/ProjectInteractorTest.java:221]: To me this one sounds sufficiently generic to promote to a constant. But I admit this is a matter of personal taste.

Comment (AnnieJovitha) [reply src/test/java/com/philips/research/bombar/core/domain/ProjectInteractorTest.java:221]: I keep it same.

File: src/test/java/com/philips/research/bombar/core/domain/DependencyTest.java [hunk 15]
```diff
@@ -138,14 +138,47 @@ void addsRelations() {
...
 
+    @Test
+    void noStrongUsageForOrphan() {
+        assertThat(dependency.getStrongestUsage()).isEmpty();
+    }
+
+    @Test
+    void findsStrongestRelationShipFromOwnUsage() {
+        final var parent = new Dependency(ID, "Parent");
+        final var firstChild = new Dependency(ID, "First");
```
Comment (timovandeput) [inline src/test/java/com/philips/research/bombar/core/domain/DependencyTest.java:156]: Isn't this about a "child" and its "sibling"? (That at least indicates the primary subject for this test.)

File: src/test/java/com/philips/research/bombar/core/domain/DependencyTest.java [hunk 16]
```diff
@@ -138,14 +138,47 @@ void addsRelations() {
...
         assertThat(dependencies).containsExactly(target);
     }
 
+    @Test
+    void noStrongUsageForOrphan() {
+        assertThat(dependency.getStrongestUsage()).isEmpty();
+    }
+
+    @Test
+    void findsStrongestRelationShipFromOwnUsage() {
```
Comment (timovandeput) [inline src/test/java/com/philips/research/bombar/core/domain/DependencyTest.java:154]: "Relationship" is a single word, not two words.

File: src/test/java/com/philips/research/bombar/core/domain/DependencyTest.java [hunk 17]
```diff
@@ -138,14 +138,47 @@ void addsRelations() {
...
+        final var parent = new Dependency(ID, "Parent");
+        final var firstChild = new Dependency(ID, "First");
+        final var secondChild = new Dependency(ID, "Second");
+        final var firstChildRelation = new Relation(Relation.Relationship.STATIC_LINK, firstChild);
+        final var secondChildRelation = new Relation(Relation.Relationship.DYNAMIC_LINK, secondChild);
+
```
Comment (timovandeput) [inline src/test/java/com/philips/research/bombar/core/domain/DependencyTest.java:160]: Why the empty line here? Why are these two sections separated?

**Review by timovandeput — APPROVED** [review]
Please check the remaining new suggestions.

File: ui/lib/services/project_service.dart [hunk 18]
```diff
@@ -119,6 +119,14 @@ class ProjectService {
     });
   }
 
+  Future<Map<String, Set<Dependency>>> projectObligation() async {
+    _assertProjectSelected();
+
+    return _execute(() async {
+      return await _client.findObligations(_currentProject!.id);
```
Comment (timovandeput) [inline ui/lib/services/project_service.dart]: The method returns a `Future`, so the combination of `return await` seems odd in an `async` method.

Please learn about Futures in Dart, and then remove the `async` and the `await` in this invocation. And since you are then left with a single method body, you can additionally change it into an "expression body".

File: ui/lib/services/project_service.dart [hunk 19]
```diff
@@ -185,7 +193,7 @@ class ProjectService {
     try {
       return await func();
     } on DioError catch (e) {
-      log('Backend communication failed', error: e.error);
+      log('Backend communication failed', error: e.error.toString());
```
Comment (timovandeput) [inline ui/lib/services/project_service.dart]: I don't think this `.toString()` is necessary. (I would expect the logger to do this internally.)

File: ui/test/services/bom_bar_client_test.dart [hunk 20]
```diff
@@ -160,6 +160,34 @@ void main() {
         });
       });
 
+      group('Find obligations for dependencies of a project', () {
+        test('finds obligations per dependency', () async {
```
Comment (timovandeput) [inline ui/test/services/bom_bar_client_test.dart]: But the server is delivering dependencies per obligation???

Correct the title to avoid confusion for maintainers of this code.

File: ui/test/services/bom_bar_client_test.dart [hunk 21]
```diff
@@ -160,6 +160,34 @@ void main() {
...
+          server.respondJson({
+            'obligation': [
```
Comment (timovandeput) [inline ui/test/services/bom_bar_client_test.dart]: Suggestion to start using constants for "'obligation'" and "`DependencyId`" to make the test more concise. (Note that you can put such constants in the scope of the group.)

Comment (timovandeput) [inline ui/test/services/bom_bar_client_test.dart]: The value is not actually an identifier-like "obligation", but it is the obligation _description_. (You might want to choose the constant name accordingly.)

File: ui/test/services/project_service_test.dart [hunk 22]
```diff
@@ -192,6 +197,19 @@ void main() {
         expect(result.keys, ['high', 'low']);
       });
 
+      test('loads obligations', () async {
+        final dependencies = <Dependency>{};
+        dependencies.add(Dependency(id: dependencyId));
```
Comment (timovandeput) [inline ui/test/services/project_service_test.dart]: This line can be combined with the previous line into:
```dart
final dependencies = {Dependency(id: dependencyId)};
```
And then you can probably also inline it into the next line, because the dependencies are not used anywhere else in the test.

File: ui/test/services/project_service_test.dart [hunk 23]
```diff
@@ -192,6 +197,19 @@ void main() {
...
+        final obligations = {'Obligation': dependencies};
+        when(client.findObligations(projectId))
+            .thenAnswer((_) => Future.value(obligations));
+
+        final result = await service.projectObligation();
+
+        expect(result, obligations);
+        expect(result['Obligation']!.first.id, dependencyId);
```
Comment (timovandeput) [inline ui/test/services/project_service_test.dart]: I suspect this lines performs a sub-check of the previous line where the total map is checked.

**Review by timovandeput — CHANGES_REQUESTED** [review]
